File: python/pages/shadowdom_page.py
from selenium.webdriver.common.by import By
from .base_page import BasePage


class ShadowDomPage(BasePage):

    HOST = (By.ID, "shadow-host")
    SHADOW_TEXT = "#shadow-text"

    def read_shadow_text(self):
        host = self.wait_visible(self.HOST)
        shadow_root = host.shadow_root
        return shadow_root.find_element(By.CSS_SELECTOR, self.SHADOW_TEXT).text

File: src/main/java/com/example/pages/ShadowDomPage.java
package com.example.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class ShadowDomPage extends BasePage {

    // Host element for Shadow DOM
    private final By shadowHost = By.cssSelector("#shadow-host");

    // Selectors *inside* shadow root
    private final By shadowText = By.cssSelector("#shadowText");
    private final By shadowBtn = By.cssSelector("#shadowBtn");
    private final String url = "http://localhost:8080/shadow-dom.html";

    public ShadowDomPage(WebDriver driver) {
        super(driver);
    }

    public ShadowDomPage load() {
        driver.get(url);
        return this;
    }

    private SearchContext getShadowRoot() {
        WebElement host = find(shadowHost);
        return (SearchContext) ((org.openqa.selenium.JavascriptExecutor) driver)
                .executeScript("return arguments[0].shadowRoot", host);
    }

    public String getShadowText() {
        return getShadowRoot().findElement(shadowText).getText();
    }

    public void clickShadowButton() {
        getShadowRoot().findElement(shadowBtn).click();
    }
}
